glob.go

```diff
@@ -12,30 +12,115 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// Package glob provides equivalent functionality to filepath.Glob while
+// meeting different performance requirements.
 package glob
 
 import (
+	"context"
+	"io"
 	"os"
 	"path/filepath"
 	"runtime"
-	"sort"
 	"strings"
 )
 
-// Glob returns the names of all files matching pattern or nil
-// if there is no matching file. The syntax of patterns is the same
-// as in filepath.Match. The pattern may describe hierarchical names such as
-// /usr/*/bin/ed (assuming the Separator is '/').
+// Glob is similar to filepath.Glob but with different performance concerns.
 //
-// Glob ignores file system errors such as I/O errors reading directories.
-// The only possible returned error is filepath.ErrBadPattern, when pattern
-// is malformed.
-func Glob(pattern string) (matches []string, err error) {
+// Firstly, It can be canceled via the context. Secondly, it makes no guarantees
+// about the order of returned matches. This change allows it to run in O(d+m)
+// memory and O(n) time, where m is the number of match results, d is the depth
+// of the directory tree the pattern is concerned with, and n is the number of
+// files in that tree.
+func Glob(ctx context.Context, pattern string) ([]string, error) {
+	gr := Stream(pattern)
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		<-ctx.Done()
+		gr.Close()
+	}()
+	defer cancel()
+
+	var ret []string
+	for {
+		match, err := gr.Next()
+		if err != nil {
+			return nil, err
+		}
+		if match == "" {
+			break
+		}
+		ret = append(ret, match)
+	}
+	return ret, nil
+}
+
+type GlobResult struct {
+	errors  chan error
+	results chan string
```

Comment:
Yeah I thought about that, but then the issue is we have to do one of:

- `stream` has to send its error down a channel instead of returning it. I actually tried writing this out and it actually looked even worse than it sounds
- `stream` still gets a `chan string` and returns its error, and then the caller wraps up all the `string`s into a `struct {string, error} {s, nil}` and forwards them, then sends `struct {string, error}{"", err}` after `stream` has returned. This doesn't seem totally insane but I'm not really sure if it overall reduces the cognitive load...

Any thoughts?